format/ComputeIR.fbs

```diff
@@ -0,0 +1,267 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+include "Schema.fbs";
+
+namespace org.apache.arrow.flatbuf.computeir;
+
+/// Avoid use of org.apache.arrow.Buffer because it requires a
+/// sidecar block of bytes.
+table InlineBuffer {
+  // ulong is used to guarantee alignment and padding of `bytes` so that flatbuffers
+  // and other alignment sensitive blobs can be stored here
+  bytes: [ulong] (required);
+}
+
+/// An expression is one of
+/// - a Literal datum
+/// - a reference to a Field from a Relation
+/// - a call to a named function
+/// On evaluation, an Expression will have either array or scalar shape.
+union ExpressionImpl {
+  Literal, FieldRef, Call
+}
+
+table Expression {
+  impl: ExpressionImpl (required);
```

Comment:
what this indirection?

Reply:
The generators for some languages don't support vector-of-unions. I'll add a comment explaining that

Reply:
Per below, if you lift the output type (field) into Expression (if we agree that every Expression needs an output Field — name and type), than you can simply call this `ExpressionOp`, since then `Expression` serves more purpose than simply being wrapper for `Expression.impl`

Reply:
I agree that every Expression needs type, but I disagree that name is meaningful for Expressions. To my mind a name is ascribed to the referent by the referring entity; to speak in graph theory it's an edge property rather than a node property. To give a practical context: in a projection like `SELECT $complicated_expr as A, $complicated_expr as B` it should not be an error to use a single memoized instance of `$complicated_expr`- but including name as an expression property would make these two incompatible.